### PR TITLE
When an invalid loc is computed, log debug info and return no code actions instead of crashing

### PR DESCRIPTION
**Pull request by neilparikh**
### Motivation

We should not crash when the user executes an extract action that triggers an edge case we don't handle.

### Test plan

Ran a sorbet release build with the `--debug-log-file=`, ran a known crashing extract action, saw that sorbet did not crash, and saw the relevant logs in the log file:
```
[2024-07-31 14:11:07.341] [consoleAndFile] [error] msg="ExtractToVariable: failed to determine whereToInsert in getExtractMultipleOccurrenceEdits" selectionLoc="Loc {file=./test.rb start=5:4 end=5:5}"
[2024-07-31 14:11:07.341] [consoleAndFile] [error] source="# typed: strict\n\na = T.unsafe([])\n\nif a && T.unsafe(1)\nend\n"
```

### CODE REVIEW

File: main/lsp/ExtractVariable.cc [hunk 1]
```diff
@@ -467,6 +480,11 @@ MultipleOccurrenceResult VariableExtractor::getExtractMultipleOccurrenceEdits(co
     auto firstMatch = matches[0];
 
     auto whereToInsert = findWhereToInsert(*scopeToInsertIn, firstMatch);
+    if (!whereToInsert.exists()) {
+        logDebugInfo(config.logger, gs, selectionLoc,
+                     "failed to determine whereToInsert in getExtractMultipleOccurrenceEdits");
+        return {};
+    }
```
Comment (jez) [inline main/lsp/ExtractVariable.cc:487]: Is it worth making `findWhereToInsert` return an `optional<Loc>` to force users to explicitly dereference the result, reminding them that they also want to look at the result?

We have a few places in `Loc.h` that use `optional` specifically for this, like `source` and `fromDetails`.

File: main/lsp/ExtractVariable.cc [hunk 2]
```diff
@@ -1,11 +1,19 @@
...
 
 using namespace std;
 
 namespace sorbet::realmain::lsp {
 
+void logDebugInfo(const std::shared_ptr<spdlog::logger> logger, const core::GlobalState &gs,
+                  const core::Loc selectionLoc, const std::string message) {
+    logger->debug("ExtractToVariable: {}", message);
+    logger->debug("selectionLoc=\"{}\"", selectionLoc.showRaw(gs));
+    logger->debug("source=\"{}\"", absl::CEscape(selectionLoc.file().data(gs).source()));
```
Comment (jez) [inline main/lsp/ExtractVariable.cc:14]: You want these to show up at `error`, not `debug`.

File: main/lsp/ExtractVariable.cc [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
 #include "ast/treemap/treemap.h"
 #include "common/sort/sort.h"
 
 using namespace std;
 
 namespace sorbet::realmain::lsp {
 
+void logDebugInfo(const std::shared_ptr<spdlog::logger> logger, const core::GlobalState &gs,
+                  const core::Loc selectionLoc, const std::string message) {
+    logger->debug("ExtractToVariable: {}", message);
```
Comment (jez) [inline main/lsp/ExtractVariable.cc:12]:
```suggestion
    logger->debug("msg=\"ExtractToVariable: {}\"", message);
```

File: main/lsp/ExtractVariable.cc [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
+    logger->debug("selectionLoc=\"{}\"", selectionLoc.showRaw(gs));
```
Comment (jez) [inline main/lsp/ExtractVariable.cc:13]: Ideally we would put all the variables that aren't `source` on one line, makes it easier to manipulate them in splunk queries. the reason why i didn't do it for `source` in the other `Loc.cc` log lines is because i was worried about my log lines getting like sampled if they're too long.

**Review by jez — APPROVED** [review]
mostly looks good but i have some suggestions.

File: main/lsp/ExtractVariable.cc (hunk 1 above)
Comment (neilparikh) [reply main/lsp/ExtractVariable.cc:486]: I don't think it's worth it, so I'll leave it as is. In theory, this call shouldn't fail; if the code has gotten this far, it should complete successfully. There's also only 2 call sites, both in this file, so I don't expect any calls to accidentally use it without checking.

Comment (jez) [reply main/lsp/ExtractVariable.cc:486]: > In theory, this call shouldn't fail

I made the Loc::source method return an optional specifically because "this should never crash, but in practice it's crashing a lot in ways that are tangential to normal execution, we shouldn't punish the user for our own faults"

#4075

so there is precedent for this, but it's ultimately up to you in this case.
